### ISSUE TITLE
"IDE0004 C# Cast is redundant" is reported incorrectly. 

### ISSUE THREAD

**Issue by thomasouvre** [dotnet/roslyn#29365]
**Version Used**: 
Microsoft Visual Studio Enterprise 2017 Version 15.8.0

**Steps to Reproduce**:
Consider the following program:
```csharp
var nullable = true;
var date = DateTime.Now;
var expr = Expression.Constant(nullable ? (DateTime?)date : date);
```

**Expected Behavior**:
The cast (DateTime?) must not be reported because the parameter type of the method Expression.Constant is an object, so the resulting expression is different depending on the cast.

**Actual Behavior**:
The cast (DateTime?) is highlighted with the following message: "Remove Unnecessary cast" and "IDE0004 C# Cast is redundant".

![ddd](https://user-images.githubusercontent.com/4851282/44269445-d330f400-a234-11e8-80c5-c46c3d53e315.png)

**Comment by thomasouvre** [comment]
Sorry it's effectively redundant, because the cast of a nullable to object looses the nullable side.
